pkg/network/tracer/connection/failure/matching.go

```diff
@@ -0,0 +1,137 @@
+	111: {}, // Connection refused
+}
+
+var telemetryModuleName = "network_tracer__tcp_failure"
+
+var failureTelemetry = struct {
+	failedConnOrphans telemetry.Counter
+	failedConnMatches telemetry.Counter
+}{
+	telemetry.NewCounter(telemetryModuleName, "orphans", []string{}, "Counter measuring the number of orphans after associating failed connections with a closed connection"),
+	telemetry.NewCounter(telemetryModuleName, "matches", []string{"type"}, "Counter measuring the number of successful matches of failed connections with closed connections"),
+}
+
+// FailedConnStats is a wrapper to help document the purpose of the underlying map
+type FailedConnStats struct {
+	CountByErrCode map[uint32]uint32
+	Expiry         int64
+}
+
+// String returns a string representation of the failedConnStats
+func (t FailedConnStats) String() string {
+	return fmt.Sprintf(
+		"FailedConnStats{CountByErrCode: %v, Expiry: %d}", t.CountByErrCode, t.Expiry,
+	)
+}
+
+// FailedConnMap is a map of connection tuples to failed connection stats
+type FailedConnMap map[ebpf.ConnTuple]*FailedConnStats
+
+// FailedConns is a struct to hold failed connections
+type FailedConns struct {
+	FailedConnMap map[ebpf.ConnTuple]*FailedConnStats
+	sync.RWMutex
+}
+
+// NewFailedConns returns a new FailedConns struct
+func NewFailedConns() *FailedConns {
+	return &FailedConns{
+		FailedConnMap: make(map[ebpf.ConnTuple]*FailedConnStats),
+	}
+}
+
+// MatchFailedConn increments the failed connection counters for a given connection based on the failed connection map
+func (fc *FailedConns) MatchFailedConn(conn *network.ConnectionStats) {
+	if conn.Type != network.TCP {
+		return
+	}
+	connTuple := connStatsToTuple(conn)
+
+	fc.RLock()
+	defer fc.RUnlock()
+
+	if failedConn, ok := fc.FailedConnMap[connTuple]; ok {
+		// found matching failed connection
+		conn.TCPFailures = make(map[uint32]uint32)
+
+		for errCode, count := range failedConn.CountByErrCode {
+			failureTelemetry.failedConnMatches.Add(1, strconv.Itoa(int(errCode)))
+			conn.TCPFailures[errCode] += count
+		}
+	}
+}
+
+// RemoveExpired removes expired failed connections from the map
+func (fc *FailedConns) RemoveExpired() {
+	if fc == nil {
+		return
+	}
+	fc.Lock()
+	defer fc.Unlock()
+
+	now := time.Now().Unix()
+	removed := 0
+
+	for connTuple, failedConn := range fc.FailedConnMap {
+		if failedConn.Expiry < now {
+			removed++
+			delete(fc.FailedConnMap, connTuple)
+		}
+	}
+
+	failureTelemetry.failedConnOrphans.Add(float64(removed))
+}
+
+// connStatsToTuple converts a ConnectionStats to a ConnTuple
+func connStatsToTuple(c *network.ConnectionStats) ebpf.ConnTuple {
+	var tup ebpf.ConnTuple
+	tup.Sport = c.SPort
+	tup.Dport = c.DPort
+	tup.Netns = c.NetNS
+	tup.Pid = c.Pid
+	if c.Family == network.AFINET {
+		tup.SetFamily(ebpf.IPv4)
+	} else {
+		tup.SetFamily(ebpf.IPv6)
+	}
+	if c.Type == network.TCP {
+		tup.SetType(ebpf.TCP)
+	} else {
+		tup.SetType(ebpf.UDP)
+	}
+	if c.Source.IsZero() {
```

Comment:
nit: you really only have to handle the `!c.Source.IsZero()` case (and also below for `c.Dest`)

Reply:
removed